src/main/java/org/truffleruby/core/module/ModuleOperations.java

```diff
@@ -397,32 +397,25 @@ public static ConstantLookupResult lookupConstantWithInherit(RubyContext context
     public static MethodLookupResult lookupMethodCached(DynamicObject module, String name,
             DeclarationContext declarationContext) {
         final ArrayList<Assumption> assumptions = new ArrayList<>();
+        DynamicObject[] refinements;
 
-        // Look in ancestors
         for (DynamicObject ancestor : Layouts.MODULE.getFields(module).ancestors()) {
-            final ModuleFields fields = Layouts.MODULE.getFields(ancestor);
-            assumptions.add(fields.getMethodsUnmodifiedAssumption());
-            final InternalMethod method = fields.getMethod(name);
-            if (method != null) {
-                if (method.isRefined()) {
-                    if (declarationContext != null) {
-                        final DynamicObject[] refinements = declarationContext.getRefinementsFor(ancestor);
-                        if (refinements != null) {
-                            for (DynamicObject refinement : refinements) {
-                                final MethodLookupResult refinedMethod = lookupMethodCached(refinement, name, null);
-                                if (refinedMethod.isDefined()) {
-                                    for (Assumption assumption : refinedMethod.getAssumptions()) {
-                                        assumptions.add(assumption);
-                                    }
-                                    return new MethodLookupResult(refinedMethod.getMethod(), toArray(assumptions));
-                                }
-                            }
+            if ((refinements = getRefinementsFor(declarationContext, ancestor)) != null) {
```

Comment:
Style thing: could you move the assignment on a separate line before, and have the declaration there too?
That way `refinements` is not reassigned which I think helps readability.

Reply:
> and have the declaration there too?

Is the variable allocation in cycle will affect performance? I specifically put a declaration outside `for` to just assignment it (btw, I'm okay with do assignment with a separate line before).

Reply:
Not, it shouldn't affect performance, at least not when the code is JIT-compiled or AOT-compiled (by Native Image using Graal). The compiler will anyway realize the variable is not used outside that branch.
See https://chrisseaton.com/truffleruby/basic-graal-graphs/#local-variables as an illustration of how Graal doesn't really mind where you declare your variables.

Reply:
I love this
![](https://thumbs.gfycat.com/FrayedAdoredBlackwidowspider-small.gif)

Corrected :)